### PR TITLE
179 status value

### LINKED ISSUE #179: Instruction can use status as value
### Content
Instruction should be able to use a status (of another device)  as value for an instruction

### Feature
- Feature of this issue

### Action plan
- [x] Add new type status
- [x] Fix type checking
- [ ] On action retrieve and send correct status
- [ ] Make an example (with controlboard and hue lights)

### LINKED ISSUE #214: bug: no confirmations sent
### Content
no confirmation messages are sent from the js client library (node implementation)

they are not logged nor received on back-end

### PR DESCRIPTION
**Pull request by mrlomar**
## Contents
Actions can now send status of devices as value

## Issue
Closes #179 
Closes #214

### CODE REVIEW

File: back-end/src/sciler/config/checkConfig.go [hunk 1]
```diff
@@ -192,12 +193,32 @@ func checkActionInstructionType(valueType reflect.Kind, instructionType string,
...
 
+// checkStatusInstructionType checks if a status is in the form: `deviceID.component`
+// and if the device and component exists
+func (config *WorkingConfig) checkStatusInstructionType(status string, ruleID string) error {
+	split := strings.Split(status, ".")
+	if len(split) == 2 {
+		deviceID := split[0]
+		componentID := split[1]
+		device := config.Devices[deviceID]
+		if device != nil && (device.Input[componentID] != "" || device.Output[componentID].Type != "") {
```
Comment (GwennanS) [inline back-end/src/sciler/config/checkConfig.go]: can you split these checks and return a specific error message

File: back-end/src/sciler/config/workingConfigTypes_test.go [hunk 2]
```diff
@@ -34,6 +34,10 @@ type HandlerMock struct {
 	mock.Mock
 }
 
+func (handlerMock *HandlerMock) PrepareMessage(typeID string, message []ComponentInstruction) []ComponentInstruction {
+	panic("implement me")
```
Comment (GwennanS) [inline back-end/src/sciler/config/workingConfigTypes_test.go]: what

File: cc_library/example_scripts_py_scc/hue_lights_config.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "id": "hue_lights",
-  "host": "localhost",
+  "host": "192.168.0.103",
```
Comment (GwennanS) [inline cc_library/example_scripts_py_scc/hue_lights_config.json]: set back to localhost plz

File: back-end/src/sciler/config/workingConfigTypes_test.go (hunk 2 above)
Comment (mrlomar) [reply back-end/src/sciler/config/workingConfigTypes_test.go]: does not need implementation

File: back-end/src/sciler/config/checkConfig.go (hunk 1 above)
Comment (mrlomar) [reply back-end/src/sciler/config/checkConfig.go]: done

File: cc_library/example_scripts_py_scc/hue_lights_config.json (hunk 3 above)
Comment (mrlomar) [reply cc_library/example_scripts_py_scc/hue_lights_config.json]: done

File: back-end/resources/manuals/room_manual.md [hunk 4]
```diff
@@ -145,7 +145,7 @@ This can be displayed in the front-end, so should be readable and in Dutch.
     - `message` in case of type `device`: this defines a list of componentInstructions which have a:
         - `component_id`: this will be the id of a component in a timer or device
         - `instruction`: one of the instructions specified for this device and component
-        - `value`: this is the value for the instruction of the type specified for this device and component
+        - `value`: this is the value for the instruction of the type specified for this device and component, in case this type is "status", a string in the form "`device_id`.`component_id`" should be specified.
```
Comment (IssaHanou) [inline back-end/resources/manuals/room_manual.md]: probably more clear as `device_id.component_id`

File: back-end/src/sciler/handler/handler_helper.go [hunk 5]
```diff
@@ -76,6 +77,25 @@ func (handler *Handler) SendComponentInstruction(clientID string, instructions [
 	}
 }
 
+// PrepareMessage scans a message and if the instruction is of type status, if so the value of the message is replaced by the status of a device
+func (handler *Handler) PrepareMessage(typeID string, messages []config.ComponentInstruction) []config.ComponentInstruction {
```
Comment (IssaHanou) [inline back-end/src/sciler/handler/handler_helper.go]: are the deviceID and componentID already checked at this stage?

File: back-end/src/sciler/config/checkConfig.go [hunk 6]
```diff
@@ -192,12 +193,35 @@ func checkActionInstructionType(valueType reflect.Kind, instructionType string,
...
+	if len(split) == 2 {
+		deviceID := split[0]
+		componentID := split[1]
+		if device := config.Devices[deviceID]; device != nil {
+			if device.Input[componentID] != "" || device.Output[componentID].Type != "" {
+				return nil
+			}
+			return fmt.Errorf("level III - implementation error: on rule %s's actions: instruction type status expected in the form of `deviceID.component` but component with ID %v not found", ruleID, componentID)
+		}
+		return fmt.Errorf("level III - implementation error: on rule %s's actions: instruction type status expected in the form of `deviceID.component` but device with ID %v not found", ruleID, deviceID)
```
Comment (IssaHanou) [inline back-end/src/sciler/config/checkConfig.go]: do the checks using `object, ok := map[objectId]; !ok` for both device and component id

File: cc_library/example_scripts_py_scc/hue_lights.py [hunk 7]
```diff
@@ -94,25 +85,60 @@ def set_scene(self, data):
...
+                    + self.hue_user
+                    + "/lights/"
+                    + comp[-1:]
+                    + "/state"
             )
+        self.pub_to_hue(url)
+
+    def set_single(self, action):
+        if action.get("instruction") == "bri":
+            # self.bri = action.get("value") * 2.5
```
Comment (IssaHanou) [inline cc_library/example_scripts_py_scc/hue_lights.py]: remove comment

Comment (IssaHanou) [reply cc_library/example_scripts_py_scc/hue_lights.py]: or set back to proper value

File: back-end/resources/manuals/room_manual.md (hunk 4 above)
Comment (mrlomar) [reply back-end/resources/manuals/room_manual.md]: dat staat er toch?

File: back-end/src/sciler/handler/handler_helper.go (hunk 5 above)
Comment (mrlomar) [reply back-end/src/sciler/handler/handler_helper.go]: yes

File: back-end/src/sciler/config/checkConfig.go (hunk 6 above)
Comment (mrlomar) [reply back-end/src/sciler/config/checkConfig.go]: why?
I will not use the object here

File: cc_library/example_scripts_py_scc/hue_lights.py (hunk 7 above)
Comment (mrlomar) [reply cc_library/example_scripts_py_scc/hue_lights.py]: will do

File: back-end/src/sciler/config/checkConfig.go (hunk 6 above)
Comment (IssaHanou) [reply back-end/src/sciler/config/checkConfig.go]: Then you can use _, OK. It's a builtin mechanism for checking whether a key exists so I'd say that is prefferee

File: back-end/resources/manuals/room_manual.md (hunk 4 above)
Comment (IssaHanou) [reply back-end/resources/manuals/room_manual.md]: Current is `device`. `component`

Comment (mrlomar) [reply back-end/resources/manuals/room_manual.md]: done

File: back-end/src/sciler/config/checkConfig.go (hunk 6 above)
Comment (mrlomar) [reply back-end/src/sciler/config/checkConfig.go]: done

**Pull request merged**